pkg/k8s/watchers/pod.go

```diff
@@ -370,6 +370,79 @@ func (k *K8sWatcher) updatePodHostIP(pod *types.Pod) (bool, error) {
 	return skipped, nil
 }
 
+// Find an endpoint corresponding to the specified pod. This function makes
+// assumptions about the specified pod being deleted, it is *NOT* a generic
+// endpoint->pod correlation function.
+//
+// Returns an Endpoint if one is found that matches the specific pod.
+func (k *K8sWatcher) findEndpointForDeletedPod(pod *types.Pod) *endpoint.Endpoint {
+	podNSName := k8sUtils.GetObjNamespaceName(&pod.ObjectMeta)
+	ep := k.endpointManager.LookupPodName(podNSName)
+	if ep == nil {
+		return nil
+	}
+
+	// We don't have a containerID for the pod, and it's not sufficient to
+	// use the pod namespace+name to uniquely identify an endpoint as the
+	// exact same pod (since a new pod can be provisioned with the same
+	// name). Instead we use the creation time of the endpoint and the
+	// start time of the pod to infer that this endpoint corresponds to the
+	// instance of the pod that is being deleted. This pattern should work
+	// for the more common case since the CNI ADD will always occur first
+	// and then the pod will start later after the network is successfully
+	// provisioned by Cilium.
+	//
+	// Account for the following cases:
+	// * Endpoint is older than the pod. If we reach this point, we know
```

Comment:
This isn't 100% reliable. The worker and the controller nodes can be 2 different machines. The creation timestamp for both entities, pod and Cilium endpoint, might be off because the clocks of both machines are not in sync.

Reply:
Hmm, so pod start time is generated at the controller not at kubelet on the node? I figured it would be populated by the node :/